p2p/net/swarm/dial_ranker.go

```diff
@@ -103,7 +104,29 @@ func getAddrDelay(addrs []ma.Multiaddr, tcpDelay time.Duration, quicDelay time.D
 						copy(addrs[2:], addrs[1:i])
 						addrs[1] = a
 					}
-					happyEyeballs = true
+					happyEyeballsQUIC = true
+					break
+				}
+			}
+		}
+		// idx is the index of the first tcp address
+		idx := 0
+		for i, a := range addrs {
+			if isProtocolAddr(a, ma.P_TCP) {
+				idx = i
+				break
+			}
+		}
+		if isProtocolAddr(addrs[idx], ma.P_TCP) && isProtocolAddr(addrs[idx], ma.P_IP6) {
+			for i := idx + 1; i < len(addrs); i++ {
+				if isProtocolAddr(addrs[i], ma.P_TCP) && isProtocolAddr(addrs[i], ma.P_IP4) {
```

Comment:
This could be done in a single loop over addrs. It would also be helpful to have a comment here to explain what this logic is doing.

Reply:
I've added comments and made a single loop. Are the comments helpful?